### PR TITLE
More Design Window fixes

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This patch is a series of fixes to bugs in the design window.  Some are related to the new replace design feature, while others pre-date that feature.  The bugs fixed are:
1. The drop indicator was active for the hulls and monster lists.  It should not be.
2. EnableOrderIssuing was not hooked up to prevent observers from changing the order of an empire's designs.
3. Dropping any hulls/designs/monsters on the workbench moved that design to the end of the list temporarily.
4. DesignNameValidity was being called too often and slowing the interface.
5. When re-ordering designs, dropping a design on itself was temporarily moving it to the end of the list.
6. Not a bug.  I upgraded error checking on QueueListBox and used its functionality more fully in BasesListBox.
7. Selecting a hull/monster was not disabling the replace button.

### CODE REVIEW

File: UI/DesignWnd.cpp [hunk 1]
```diff
@@ -1484,7 +1486,11 @@ void BasesListBox::ChildrenDraggedAway(const std::vector<GG::Wnd*>& wnds, const
     const GG::Control* control = dynamic_cast<const GG::Control*>(wnd);
     if (!control)
         return;
-    DetachChild(wnds.front());
+
+    Row* original_row = boost::polymorphic_downcast<Row*>(*wnds.begin());
+    iterator insertion_point = std::find(begin(), end(), original_row);
+    if(insertion_point != end())
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: space between if and (

File: UI/DesignWnd.cpp [hunk 2]
```diff
@@ -1533,69 +1528,38 @@ void BasesListBox::ChildrenDraggedAway(const std::vector<GG::Wnd*>& wnds, const
...
     CUIListBox::ChildrenDraggedAway(wnds, destination);
     DetachChild(wnds.front());
-
 }
 
+void BasesListBox::QueueItemMoved(GG::ListBox::Row* row, std::size_t position) {
+    BasesListBox::CompletedDesignListBoxRow* control =
+        boost::polymorphic_downcast<BasesListBox::CompletedDesignListBoxRow*>(row);
+    Empire* empire = GetEmpire(m_empire_id_shown);
+    if (control != NULL && empire != NULL) {
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: avoid NULL, just do

```
if (control && empire)
```

File: UI/DesignWnd.cpp [hunk 3]
```diff
@@ -1533,69 +1528,38 @@ void BasesListBox::ChildrenDraggedAway(const std::vector<GG::Wnd*>& wnds, const
...
-    if (std::distance(first, last) != 1) {
-        ErrorLogger() << "SavedDesignPanel::DropsAcceptable expects a single dropped item.";
-        return;
-    }
-
-    if (m_showing_completed_designs && first->first->DragDropDataType() == COMPLETE_DESIGN_ROW_DROP_STRING)
-        first->second = true;
-}
+        iterator insert_before_row = begin();
+        for(std::size_t ii = 0; ii < position; ++ii)
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: space

File: UI/DesignWnd.cpp [hunk 4]
```diff
@@ -2050,6 +2018,12 @@ void BasesListBox::HideAvailability(bool available, bool refresh_list) {
     }
 }
 
+void BasesListBox::EnableOrderIssuing(bool enable/* = true*/)
+{
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: move { to previous line

File: UI/DesignWnd.cpp [hunk 5]
```diff
@@ -2215,6 +2190,17 @@ void DesignWnd::BaseSelector::ToggleAvailability(bool available, bool refresh_li
     }
 }
 
+void DesignWnd::BaseSelector::EnableOrderIssuing(bool enable/* = true*/) {
+    if(m_hulls_list)
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: spaces...

File: UI/DesignWnd.cpp [hunk 6]
```diff
@@ -3497,4 +3483,6 @@ void DesignWnd::ReplaceDesign() {
 }
 
 void DesignWnd::EnableOrderIssuing(bool enable/* = true*/)
-{}
+{
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: put on one line

File: UI/QueueListBox.cpp [hunk 7]
```diff
@@ -112,8 +116,12 @@ void QueueListBox::DropsAcceptable(DropsAcceptableIter first, DropsAcceptableIte
 }
 
 void QueueListBox::AcceptDrops(const GG::Pt& pt, const std::vector<GG::Wnd*>& wnds, GG::Flags<GG::ModKey> mod_keys) {
-    if (wnds.size() != 1)
-        return;
+    if (wnds.size() != 1) {
```
Comment (geoffthemedio) [inline UI/QueueListBox.cpp]: probably should be if wnds.size() > 1

File: UI/DesignWnd.cpp [hunk 8]
```diff
@@ -1533,69 +1528,38 @@ void BasesListBox::ChildrenDraggedAway(const std::vector<GG::Wnd*>& wnds, const
...
-        if (control && empire) {
-            int design_id = control->DesignID();
-
-            iterator insert_before_row = RowUnderPt(pt);
-            const BasesListBox::CompletedDesignListBoxRow* insert_before_control =
-                boost::polymorphic_downcast<const BasesListBox::CompletedDesignListBoxRow*>(*insert_before_row);
-            int insert_before_id = (insert_before_row == end() || !insert_before_control)
-                ? ShipDesign::INVALID_DESIGN_ID : insert_before_control->DesignID();
+        iterator insert_before_row = begin();
+        for (std::size_t ii = 0; ii < position; ++ii)
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: prefer std::advance perhaps?

File: UI/DesignWnd.cpp [hunk 9]
```diff
@@ -1533,69 +1528,38 @@ void BasesListBox::ChildrenDraggedAway(const std::vector<GG::Wnd*>& wnds, const
...
+            insert_before_row++;
+        const BasesListBox::CompletedDesignListBoxRow* insert_before_control =
+            boost::polymorphic_downcast<const BasesListBox::CompletedDesignListBoxRow*>(*insert_before_row);
+        int insert_before_id = (insert_before_row == end() || !insert_before_control)
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: should check if an iterator is end() before dereferencing it

File: UI/DesignWnd.cpp [hunk 10]
```diff
@@ -2215,6 +2189,17 @@ void DesignWnd::BaseSelector::ToggleAvailability(bool available, bool refresh_li
     }
 }
 
+void DesignWnd::BaseSelector::EnableOrderIssuing(bool enable/* = true*/) {
+    if (m_hulls_list)
+        m_hulls_list->EnableOrderIssuing(enable);
+    if(m_designs_list)
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: spaces...

File: UI/DesignWnd.cpp [hunk 11]
```diff
@@ -3496,5 +3481,6 @@ void DesignWnd::ReplaceDesign() {
     DebugLogger() << "Replaced design #" << replaced_id << " with #" << new_design_id ;
 }
 
-void DesignWnd::EnableOrderIssuing(bool enable/* = true*/)
-{}
+void DesignWnd::EnableOrderIssuing(bool enable/* = true*/) {
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: put on one line
{ Code(); }

File: UI/QueueListBox.cpp [hunk 12]
```diff
@@ -112,17 +116,23 @@ void QueueListBox::DropsAcceptable(DropsAcceptableIter first, DropsAcceptableIte
...
     const std::string& drop_type = wnd->DragDropDataType();
-    if (AllowedDropTypes().find(drop_type) == AllowedDropTypes().end())
-        return;
     GG::ListBox::Row* row = boost::polymorphic_downcast<GG::ListBox::Row*>(wnd);
-    if (!row)
-        return;
-    if (std::find(begin(), end(), row) == end())
+    if (AllowedDropTypes().find(drop_type) == AllowedDropTypes().end() ||
+        !row ||
+        std::find(begin(), end(), row) == end()) {
```
Comment (geoffthemedio) [inline UI/QueueListBox.cpp]: move { to newline after split-row logical expression in if (...)

**Comment by geoffthemedio** [comment]
Not sure about DesignNameValidity calling, as it now doesn't update the incomplete design pedia entry until you manipulate a part, as far as I saw...

**Comment by LGM-Doyle** [comment]
I agree it only updates the pedia when a part is selected, which was the old behavior before my first patch.

However, this patch fixes the specific problem that in the late game checking the design validity could not keep up with typing speed.  The interface was sluggish to the point of being broken.

One thing that I still find annoying, is that clicking on the DesignDetails panel updates the pedia with the ship hull and not the incomplete design.  The only way to get the incomplete design into the pedia is to add (or remove and add if the design if full) a part.

**Pull request merged**